Zend/Optimizer/zend_optimizer.c

```diff
@@ -831,6 +831,57 @@ zend_class_entry *zend_optimizer_get_class_entry_from_op1(
 	return NULL;
 }
 
+const zend_class_constant *zend_fetch_class_const_info(const zend_script *script, const zend_op_array *op_array, const zend_op *opline)
```

Comment:
I think it would be good to indicate whether the result is a possible prototype. This avoids re-checking op1 again in the caller. The final check can then be moved here.

Reply:
Yes, it's a good idea to rename it, but the final check cannot be simply moved to the function because a final class constant is only needed by pass 1, inference needs a typed class constant when the constant is referenced through `static::` or `parent::`.

So I could add a new param like `fetched_class_const_type` (or similar with a more expressive name) which accepts `CLASS_CONST_TYPED` or `CLASS_CONST_FINAL`. WDYT? Do you have a better idea?

Reply:
What I mean is just adding a bool pointer that is set depending on whether class constant is guaranteed to be this exact one, or possibly one of a subclass. It will always be this one, except when it was accessed on static and the class is not final. This implicitly moves the final check here. You can then just chech the prototype param in the caller.

Reply:
OK, I see it now